### PR TITLE
add GigaSpeech dataset in SpeechBrain

### PR DESCRIPTION
**Pull request by Adel-Moumen**
# Titouan took over and edited the crap out of this post. (Thanks Adel for the amazing initial work). 

## What does this PR do?

This PR adds a first SSL fine-tuning (CTC only) recipe for ASR for GigaSpeech and a streaming conformer. Another PR will be added later to add Conformer AED (depending on GPU availability?)
## Dataset compliance

We can work with the original data or HuggingFace. We recommend using HF as the download speed is immensely faster.
 
## General Todo
To do:
- [x] Write the recipe
- [x] Train on XL (ongoing), must be done in parallel by @asumagic so we can update the recipe later with the model.

## References

k2 Icefall
Model  | Dev | Test
conformer_ctc | 10.47 | 10.58
pruned_transducer_stateless2 | 10.40 | 10.51

ESPnet
Conformer-AED | 10.90 | 10.80

### CODE REVIEW

**Comment by TParcollet** [comment]
BON, what must be done here?

File: recipes/GigaSpeech/ASR/CTC/README.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+to do
+
+```bash
+mkdir lm
+git clone https://huggingface.co/wgb14/gigaspeech_lm lm
```
Comment (asumagic) [inline recipes/GigaSpeech/ASR/CTC/README.md:5]: fyi this is missing git LFS initialization, the `.gz` files are otherwise empty

Comment (Adel-Moumen) [reply recipes/GigaSpeech/ASR/CTC/README.md:5]: I mean ofc you need to pull xD

File: recipes/GigaSpeech/ASR/CTC/README.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
...
+gunzip -c lm/3gram_pruned_1e7.arpa.gz > lm/3gram_pruned_1e7.arpa
+gunzip -c lm/4gram.arpa.gz > lm/4gram.arpa
```
Comment (asumagic) [inline recipes/GigaSpeech/ASR/CTC/README.md:7]:
```suggestion
cd lm
wget https://huggingface.co/wgb14/gigaspeech_lm/resolve/main/3gram_pruned_1e7.arpa.gz
wget https://huggingface.co/wgb14/gigaspeech_lm/resolve/main/4gram.arpa.gz
gunzip -c 3gram_pruned_1e7.arpa.gz > 3gram_pruned_1e7.arpa
gunzip -c 4gram.arpa.gz > 4gram.arpa
```
also works

**Comment by TParcollet** [comment]
Taking over this. Training models. We will just release an SSL fine-tuning **at first**. A conformer Transducer or AED will follow. I will merge before we have the model, @asumagic will need to train one as well following the best yaml I provide to replicate. It's a wavlmlarge.

File: recipes/GigaSpeech/ASR/transducer/train.py [hunk 3]
```diff
@@ -0,0 +1,539 @@
...
+
+    # Use torchaudio if the device is CPU
+    if run_opts.get("device") == "cpu":
+        if "use_torchaudio: True" in overrides:
+            overrides.replace("use_torchaudio: True", "use_torchaudio: False")
+        else:
+            overrides += "\nuse_torchaudio: True"
+
+    # create ddp_group with the right communication protocol
+    sb.utils.distributed.ddp_init_group(run_opts)
```
Comment (Adel-Moumen) [inline recipes/GigaSpeech/ASR/transducer/train.py:427]: why is the DDP group init here and not latter ?

File: recipes/GigaSpeech/ASR/transducer/train.py [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+            "output_dev": hparams["valid_csv"],
+            "output_test": hparams["test_csv"],
+            "json_file": hparams["json_file"],
+            "skip_prep": hparams["skip_prep"],
+            "convert_opus_to_wav": hparams["convert_opus_to_wav"],
+            "download_with_HF": hparams["download_with_HF"],
+            "punctuation": hparams["keep_punctuation"],
+            "filler": hparams["keep_filler_words"],
+        },
+    )
```
Comment (Adel-Moumen) [inline recipes/GigaSpeech/ASR/transducer/train.py:459]: actually I am wondering about this. What happens if all the proc are doing the data prep at the same time? Does it mens they will all download the datasets (and will override the downloaded files?)?

File: recipes/GigaSpeech/gigaspeech_prepare.py [hunk 5]
```diff
@@ -0,0 +1,693 @@
...
+        )
+        logger.info(
+            "To change this directory modify the HF_HUB_CACHE env. variable."
+        )
+
+        nproc = 1
+        if hf_multiprocess_load:
+            import multiprocessing
+
+            nproc = multiprocessing.cpu_count()
```
Comment (Adel-Moumen) [inline recipes/GigaSpeech/gigaspeech_prepare.py:221]: isn't `nproc` supposed to be `num_workers` ?

**Review by Adel-Moumen — COMMENTED** [review]
Hi, 

Thanks @TParcollet for working on this important recipe! I left some comments :) I will try to run the recipe on the smallest subset ASAP.

File: recipes/GigaSpeech/ASR/CTC/hparams/train_hf_wavlm.yaml [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+character_coverage: 1.0
+
+# Model parameters
+dnn_neurons: 1024
+dropout: 0.1
+freeze_wav2vec: False
+wav2vec_output_dim: 1024
+
+# Outputs
+output_neurons: 29 # without punctuation
```
[the remark this answers is not in the document]
Comment (TParcollet) [reply recipes/GigaSpeech/ASR/CTC/hparams/train_hf_wavlm.yaml:100]: Yes, but actually only on the S, I assumed it would be correct for the XL as well. I think that is true tbh as I used a new text normalisation.

File: recipes/GigaSpeech/ASR/transducer/train.py [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+            best_hyps, scores, _, _ = self.hparams.Greedysearcher(x)
+            return logits_transducer, wav_lens, best_hyps
+        else:
+            (
+                best_hyps,
+                best_scores,
+                nbest_hyps,
+                nbest_scores,
+            ) = self.hparams.Beamsearcher(x)
+            return logits_transducer, wav_lens, best_hyps
```
[the remark this answers is not in the document]
Comment (TParcollet) [reply recipes/GigaSpeech/ASR/transducer/train.py:147]: Agreed. Dunno how to do this well.

File: recipes/GigaSpeech/ASR/transducer/train.py (hunk 3 above)
Comment (TParcollet) [reply recipes/GigaSpeech/ASR/transducer/train.py:427]: This is done in purpose. If someone starts the recipe with DDP, we don't want it to be initialised before calling the dataprep. If this same person uses the XL or L sets, the dataprep time will end up generating a DDP timeout. This way, there is ALWAYS only one thread entering the data prep. I kept the run_on_main decoration in case our user just starts modifying the code and putting weird stuff before.

File: recipes/GigaSpeech/ASR/transducer/train.py (hunk 4 above)
Comment (TParcollet) [reply recipes/GigaSpeech/ASR/transducer/train.py:459]: It will crash due to concurrent accesses imho. Data prep is already multi processed by HF, no need to do it on our side.

File: recipes/GigaSpeech/gigaspeech_prepare.py (hunk 5 above)
Comment (TParcollet) [reply recipes/GigaSpeech/gigaspeech_prepare.py:221]: No. nproc here is independent to torch. It's the actual number of available cores, to make it faster.

File: recipes/GigaSpeech/gigaspeech_prepare.py [hunk 8]
```diff
@@ -0,0 +1,692 @@
...
+
+
+def create_csv(
+    csv_file: str,
+    info: json,
+    data_folder: str,
+    split: str,
+    convert_opus_to_wav: bool,
+    punctuation: bool = False,
+    filler: bool = False,
```
Comment (asumagic) [inline recipes/GigaSpeech/gigaspeech_prepare.py:323]: those are never passed to process_line, same for the other create_csv func

File: recipes/GigaSpeech/gigaspeech_prepare.py [hunk 9]
```diff
@@ -0,0 +1,703 @@
...
+    Notes
+    -----
+    The function iterates over predefined garbage utterance tags (GARBAGE_UTTERANCE_TAGS)
+    and removes them from the input text. It then iterates over predefined punctuation tags
+    (PUNCTUATION_TAGS) and replaces them with the corresponding punctuation.
+
+    Examples
+    --------
+    >>> text = " DOUGLAS MCGRAY IS GOING TO BE OUR GUIDE YOU WALK THROUGH THE DOOR <COMMA> YOU SEE THE RED CARPETING <COMMA> YOU SEE SOMEONE IN A SUIT <PERIOD> THEY MAY BE GREETING YOU <PERIOD>"
+    >>> preprocess_text(text)
```
Comment (asumagic) [inline recipes/GigaSpeech/gigaspeech_prepare.py:626]: Example isn't passing enough arguments

File: recipes/GigaSpeech/gigaspeech_prepare.py [hunk 10]
```diff
@@ -0,0 +1,703 @@
...
+    "douglas mcgray is going to be our guide you walk through the door, you see the red carpeting, you see someone in a suit. they may be greeting you."
```
Comment (asumagic) [inline recipes/GigaSpeech/gigaspeech_prepare.py:627]: Example output is incorrect

File: recipes/GigaSpeech/gigaspeech_prepare.py [hunk 11]
```diff
@@ -0,0 +1,703 @@
...
+    Returns
+    -------
+    None
+    """
+    total_duration = 0.0
+    nb_samples = 0
+
+    to_remove = GARBAGE_UTTERANCE_TAGS
+    if not punctuation:
+        to_remove += PUNCTUATION_TAGS
```
Comment (asumagic) [inline recipes/GigaSpeech/gigaspeech_prepare.py:363]: Not correct because this is a dict. Shouldn't this be removed because of how this is handled in `preprocess_text`, anyway?

File: recipes/GigaSpeech/gigaspeech_prepare.py (hunk 10 above)
Comment (asumagic) [inline recipes/GigaSpeech/gigaspeech_prepare.py:627]: Changing the example to demonstrate `punctuation=True`, since otherwise this would not actually change anything in the input. (Maybe should just pick a better example)

```suggestion
    >>> preprocess_text(text, punctuation=True, stopwords=GARBAGE_UTTERANCE_TAGS)
    "DOUGLAS MCGRAY IS GOING TO BE OUR GUIDE YOU WALK THROUGH THE DOOR, YOU SEE THE RED CARPETING, YOU SEE SOMEONE IN A SUIT. THEY MAY BE GREETING YOU."
```

File: recipes/GigaSpeech/gigaspeech_prepare.py (hunk 8 above)
Comment (asumagic) [reply recipes/GigaSpeech/gigaspeech_prepare.py:332]: Handled by to_remove, this is fine now (if it wasn't already?)

**Review by asumagic — APPROVED** [review]
LGTM. Training for transducer worked once and evaluated correctly, but retraining with 500k steps, batch len of 500 on 4x A100 80GB
